refactor: generic function argument

johnhooks opened the pull request:
The current export of the types as `export =` is troublesome for importing the module without enabling synthetic defaults, which isn't aways an option.

- Refactor the `memize` function to be copy the parameters and return type of the input function to the output memized function.

- Updated a bunch of deps, added a module export, and CommonJS build output,

johnhooks committed:
refactor: genertic function argument

johnhooks commented:
@aduth I tried to not change so much... but it was hard. The only part that doesn't seem to be working is the benchmark script. Let me know what you think.

Edit: I change it to`"type": "module"`, it seemed simpler, and the declaration file was output with an `m` in it when made from an `.mjs` file.

aduth commented on .eslintrc.json:
```diff
@@ -1,6 +1,6 @@
 {
 	"root": true,
-	"extends": "@aduth/eslint-config/es5",
```
Curious why this change was needed?

johnhooks replied:
I was trying to find some issue with the ESlint config, and this was changed and I didn't change it back, I'll fix it.

johnhooks replied:
Okay, so at the moment I changed this all to be an ES Module, so we wouldn't want es5 right? There is a CommonJS build artifact for supporting that use case, does that create a breaking change?

aduth replied:
Historically I'd sometimes made a point to target ES5 with the exception of ESM import/export, since most bundlers would handle the import semantics, but wanting broad compatibility for the rest of the code ([old Rollup docs used to suggest this](https://github.com/rollup/rollup/wiki/pkg.module#wait-it-just-means-import-and-export--not-other-future-javascript-features)). But I think it's safe to say that IE11 is finally dead at this point, so it's probably fine to move on from ES5. I think if we make the other breaking change with dropping CJS output, it could make sense to include a note that official support for ES5 is being dropped as well.

aduth commented on LICENSE.md:
Was it intentional to delete this file? I think it should remain.

johnhooks replied:
I'm sorry, it was not. I will fix that.

aduth commented on index.js:
```diff
@@ -26,6 +26,13 @@
+/**
+ * The memoized function.
+ *
+ * @template {(...args: unknown[]) => unknown} F
```
Wonder if it makes sense to pull out this type to share between the `@template` here and in the `memize` signature as well.

johnhooks replied:
What do you think of this?

```js
/**
 * Accepts a function to be memoized, and returns a new memoized function, with
 * optional options.
 *
 * @template {(...args: any[]) => any} F
 *
 * @param {F}             fn        Function to memoize.
 * @param {MemizeOptions} [options] Options object.
 *
 * @return {((...args: Parameters<F>) => ReturnType<F>) & MemizeMemoizedFunction} Memoized function.
 */
 ```

aduth commented on test/index.js:
```diff
@@ -1,6 +1,6 @@
-const assert = require( 'assert' );
-const sinon = require( 'sinon' );
-const memoize = require( '../' );
+import assert from 'assert'
+import sinon from 'sinon'
+import memoize from '../index.js'
```
I suspect ESLint may not be testing this file, since I'd expect semicolons.

```suggestion
import assert from 'assert';
import sinon from 'sinon';
import memoize from '../index.js';
```

johnhooks replied:
Is the rule of semicolons something that's typically enforced by `prettier` in your ESLint config. I can add `prettier` in, but right now I don't see ESLint errors about the semicolons

aduth reviewed (commented):
Hi again @johnhooks ! Thanks for the pull request.

Would it make the changes any easier if the package were ported fully to ES Modules, and/or if the pre-built UMD file were removed? These would be breaking changes, but I'd be open to them.

johnhooks committed:
chore: apply @aduth's suggestions

johnhooks commented:
@aduth I'm sorry this took a lot of tweaking to get working, and I didn't do a through review of the changes. I'll fix all this up, remove all the unintended edits.

johnhooks committed:
chore: walk back some changes

johnhooks commented on .eslintrc.json:
```diff
@@ -1,16 +1,19 @@
 	},
 	"rules": {
+		"no-var": "off",
 		"one-var": "off",
+		"prefer-spread": "off",
 		"jsdoc/check-tag-names": "off",
-		"valid-jsdoc": "off"
+		"valid-jsdoc": "off",
+		"no-redeclare": "off"
```
Do you want to disable these, or update the code to fix the errors?

aduth replied:
Was the issue because it's using the non-TypeScript version of the rule? I thought I'd seen that in an earlier commit. If it's a case of the TypeScript version working better, that sounds fine to enable the TypeScript version if the code passes with it. (The common ESLint configuration should probably be doing that by default)

johnhooks replied:
I disabled the rules to make the existing code pass. Should I modify the existing code to use `let` and `const` over `var` and use a spread of arguments instead of `apply`?

johnhooks commented on .gitignore:
```diff
@@ -3,4 +3,3 @@ node_modules
 
 # Generated
 dist
-index.d.ts
```
I moved the type declaration to the `dist` folder because that's where all the build assets are, I don't have a strong opinion on this, it just makes sense to me.

aduth replied:
Yeah, that seems like a better place for it 👍

johnhooks commented on package.json:
```diff
@@ -2,24 +2,25 @@
   "version": "1.1.0",
   "description": "Unabashedly-barebones memoization library with an aim toward speed",
-  "main": "index.js",
+  "main": "dist/index.cjs",
+  "module": "dist/index.mjs",
+  "types": "dist/index.d.ts",
+  "type": "module",
```
This should make the package support usage as either a ES Module or CommonJS correct?

johnhooks commented on rollup.config.js:
```diff
@@ -1,20 +1,39 @@
-		'file': __dirname + '/dist/memize.js',
-		format: 'iife',
-		name: 'memize',
-	},
```
I removed this format, though if you want to keep supporting it I can add it back in, it really doesn't take much to build all the different formats.

aduth replied:
Nah, I think it's fine to remove this output format. I'd probably consider it a breaking change though.

johnhooks commented on rollup.config.js:
```diff
@@ -1,20 +1,39 @@
-		replace( {
-			preventAssignment: true,
-			values: {
-				'process.env.NODE_ENV': JSON.stringify( process.env.NODE_ENV )
```
I didn't apply this setting to the CommonJS build, should it be?

aduth replied:
I think we probably don't need the `replace` at all anymore. IIRC, it was useful for the pre-built, minified output, so that Uglify/Terser would drop a condition that looked like...


```js
if ('production' === 'test') {
  // ...
}
```

([related code](https://github.com/aduth/memize/blob/6866cf6fd5c9e087c3ddea6886a1af1471bb9052/index.js#L150))

We could probably remove the `NODE_ENV=production` part of the `build:bundle` script in `package.json` as well, since it was related to this.

johnhooks commented on test/index.js:
```diff
@@ -1,6 +1,6 @@
-const memoize = require( '../' );
+import assert from 'assert';
+import sinon from 'sinon';
+import memoize from '../index'
```
This was just a test, I'm not getting any ESLint errors about semis anywhere, I tried adding `prettier` as a dep and a `.prettierrc`, but it's still not working.

johnhooks commented on tsconfig.json:
```diff
@@ -1,9 +1,9 @@
 	"compilerOptions": {
 		"baseUrl": ".",
-		"module": "CommonJS",
+		"target": "ES2017",
+		"lib": ["ES2017", "DOM", "DOM.Iterable"],
```
These are the same settings used in `hpq`

johnhooks commented:
@aduth this is in a better state for review. 

> Would it make the changes any easier if the package were ported fully to ES Modules, and/or if the pre-built UMD file were removed? These would be breaking changes, but I'd be open to them.

I moved it to ES Modules, and did remove the UMD builds. If that's not the direction you want to go, please let me know and I can revert it.  I don't know exactly the purpose the the call to `process.env` I added the replace plugin for the `es` build, but left it out on the `cjs` build. please let me know what the appropriate settings would be.

> I suspect ESLint may not be testing this file, since I'd expect semicolons.

I'm having trouble figuring out why ESLint isn't flagging semicolons. Could you take a look at it and see how your custom ESLint config is being applied and what we need to do to fix the issue?

**Additionally**

- I walked back some package updates to see if it fixed the ESLint issue, but it didn't. I could update everything and chase down the issues. This would be easier for me than to figuring out the config of old versions of the dependencies.

aduth commented on package.json:
```diff
@@ -2,25 +2,24 @@
   "version": "1.1.0",
   "description": "Unabashedly-barebones memoization library with an aim toward speed",
-  "main": "index.js",
+  "main": "dist/index.cjs",
+  "module": "dist/index.js",
```
I'm curious how well `module` is supported outside Webpack & Rollup for things like tree-shaking, and wonder if it might be better to...

1. Drop the `cjs` output altogether and require ESM (breaking change, but that's okay I think?)
2. Or, use the [`package.json` `exports` semantics](https://nodejs.org/api/packages.html#dual-commonjses-module-packages) for better future-compatibility

johnhooks replied:
I can drop type module. Though if we don't do a cjs version, can it be loaded by node when it uses `export default` syntax?

johnhooks replied:
I'm getting errors when running the tests of needing `.mjs` extensions or to use type `module`. 

```
(node:24777) Warning: To load an ES module, set "type": "module" in the package.json or use the .mjs extension.
(Use `node --trace-warnings ...` to show where the warning was created)
```

When I added the `.mjs` extension to the root `index.js` file with all the `memize` source code it in to `index.mjs` the type declaration ended up being named `index.d.mts` which seems odd.

johnhooks commented:
> We could probably remove the NODE_ENV=production part of the build:bundle script in package.json as well, since it was related to this.

We still need to remove the `process.env` otherwise it will cause an error in the browser for accessing a property on the undefined `process` variable.

> Was the issue because it's using the non-TypeScript version of the rule? I thought I'd seen that in an earlier commit. If it's a case of the TypeScript version working better, that sounds fine to enable the TypeScript version if the code passes with it. (The common ESLint configuration should probably be doing that by default)

The TypeScript version didn't make a difference, it was part of how I was trying to track down the issue. 

I don't know if the rules of your default config are being applied, when I add `"semi": "error"` to the rules, it then errors on missing semicolons.

johnhooks committed:
chore: remove type module from package.json

johnhooks commented:
While testing this out with in Gutenberg this error happened. Changing `unknown` to `any` fixed it `(...args: any[]) => any`.

```
packages/core-data/src/hooks/use-query-select.ts:101:34 - error TS2345: Argument of type 'EnrichedSelectors' is not assignable to parameter of type '(...args: unknown[]) => unknown'.
  Types of parameters 'selectors' and 'args' are incompatible.
    Type 'unknown' is not assignable to type 'Record<string, (...args: any[]) => any>'.

101 const enrichSelectors = memoize( ( ( selectors ) => {
                                     ~~~~~~~~~~~~~~~~~~~~
102  const resolvers = {};
    ~~~~~~~~~~~~~~~~~~~~~~
... 
140  return resolvers;
    ~~~~~~~~~~~~~~~~~~
141 } ) as EnrichedSelectors );
    ~~~~~~~~~~~~~~~~~~~~~~~~
```

FYI: It was also good to know the original issue I was trying to solve is fixed.

johnhooks committed:
refactor: use any rather than unknown

aduth committed:
Upgrade lockfile version 3

aduth committed:
Revert to ES5 config with ecmaVersion for ESM support

aduth committed:
Remove Prettier

aduth committed:
Use type module for JS extension

aduth committed:
Enforce ESLint for benchmark and rollup config

aduth committed:
Add CHANGELOG

aduth committed:
Remove Babel

aduth committed:
Change MJS to JS

aduth committed:
Fix lint error

aduth commented:
@johnhooks Hopefully you don't mind me pushing directly to your branch. I just pushed up some commits. Would you mind taking a look over them and letting me know if you'd have any objections or if I overstepped at all?

A few remarks:

- I think we probably could move toward the non-ES5 version of the ESLint configuration, but there's a lingering question of what's a reasonable baseline support for syntax that I'm not really sure how to address. I think changing the `ecmaVersion` to support ESM and leaving the rest alone for now may help reduce the scope.
   - I think this also makes it safe to avoid Babel, since the source syntax is already targeting broad support, and we are no longer compiling for CommonJS
- I like the idea of introducing Prettier, but it'll have a pretty widespread impact on the source, more than I'd want to chew off here. I think the only thing that had been missing from what you added was the `eslint-plugin-prettier` and `eslint-config-prettier` dependencies that the shared ESLint configuration [keys off of](https://github.com/aduth/eslint-config#variations)
   - This is probably why semicolons aren't being enforced, since the shared configuration probably doesn't bother much with styling, and assumes we'd enable Prettier if we cared about it.
- With `"type": "module"` we should be able to avoid the `.mjs` changes
- Good point about needing to keep the `NODE_ENV` replacement, since it's not guaranteed that a consuming project would be replacing that in its browser builds

aduth committed:
Simplify diff, use consistent spacing style

aduth committed:
Simplify diff

johnhooks commented on rollup.config.js:
```diff
@@ -1,12 +1,11 @@
-		'file': __dirname + '/dist/memize.js',
-		format: 'iife',
+		format: 'es',
 		name: 'memize',
```
I wasn't sure about this, if outputting to `dist/index.js` does the name properly do anything?

aduth replied:
Yeah, that's a good call. I think it was only needed for the IIFE version, so we can drop it now.

johnhooks replied:
That was just a comment, do you want me to remove it?

aduth replied:
Removed it in 0745bfb

johnhooks reviewed (commented):
@aduth this looks great to me and it works beautifully in Gutenberg. It should resolve a few PRs related to improving types. I made a small comment about the `name` in the rollup config, does it have a purpose now that it being output to `dist/index.js`? 

My reason for changing it was to keep the name the same as the types export, it doesn't really matter but it makes sense to me for them to be named the same.

aduth committed:
Remove unneeded Rollup name

See https://github.com/aduth/memize/pull/10#discussion_r1177183961

aduth commented:
Thanks for another great pull request! I'll try to get this published shortly.

johnhooks commented:
Thank you @aduth, always a pleasure!

aduth commented:
This is published now as `memize@2.0.0` 🚀

johnhooks commented:
That's awesome!
